### PR TITLE
WT-2382 Resolve 'u' vs. 'U' format mismatches for cursor joins in collators/extractors.

### PR DESCRIPTION
**Pull request by ddanderson**
This includes fixes:

- resolved "'u' at end of format" by forcing 'u' to be used always.
- fixed some logic around when repacking needs to be done.
- fixed error path in __curjoin_extract_insert.
- when the first index is just 'eq', finish iteration when it becomes
  non-equal.

In additions, some cleanups related to these fixes, see commit log for more details.

To test collator/extractors, extended test_join03 to use collators.  It needs the sample collator in pull request #2473 (WT-2375) to run.  For now, the collator part of the test is disabled.

This runs successfully against the test program in WT-2382.

### CODE REVIEW

**Comment by ddanderson** [comment]
@michaelcahill, this is ready to review.  This supersedes the work in pull request #2460.

File: src/cursor/cur_join.c [hunk 1]
```diff
@@ -313,18 +313,15 @@ __curjoin_init_bloom(WT_SESSION_IMPL *session, WT_CURSOR_JOIN *cjoin,
...
-				    &allocbuf));
-			} else
-				curkey = cindex->child->key;
+			/*
+			 * Repack so it's comparable to the
+			 * reference endpoints.
+			 */
+			WT_ERR(__wt_struct_repack(session,
+			    cindex->child->key_format,
+			    (entry->repack_format ? entry->repack_format :
```
Comment (michaelcahill) [inline src/cursor/cur_join.c]: The ternary condition should be a boolean, our style is to be explicit so this should be `entry->repack_format != NULL ? ...`

File: src/cursor/cur_join.c [hunk 2]
```diff
@@ -378,19 +374,16 @@ __curjoin_endpoint_init_key(WT_SESSION_IMPL *session,
...
 		if (entry->index != NULL) {
 			/* Extract and save the index's logical key. */
 			cindex = (WT_CURSOR_INDEX *)endpoint->cursor;
 			WT_ERR(__wt_struct_repack(session,
 			    cindex->child->key_format,
-			    cindex->iface.key_format,
-			    &cindex->child->key, &endpoint->key, &allocbuf));
-			if (allocbuf != NULL)
-				F_SET(endpoint, WT_CURJOIN_END_OWN_KEY);
+			    (entry->repack_format ? entry->repack_format :
```
Comment (michaelcahill) [inline src/cursor/cur_join.c]: similarly, check against `NULL`.

**Comment by michaelcahill** [comment]
@ddanderson, thanks, apart from the minor style issues, this looks good to me.  Feel free to merge once those issues are taken care of.

**Pull request merged**